[dyn-alloc-lk] Lightning Kokkos support dynamic qubit allocation

josephleekl opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
This PR adds support for dynamic allocation for LK, and merges into [dyn-alloc-base], which contains the update for dynamic allocation for lightning.qubit. 

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-82704]

AmintorDusko commented:
Your CI, [Testing without binary-GPU](https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/12999535417/job/36255045493?pr=1049), is currently failing. Could you please check that?

josephleekl commented:
> Your CI, [Testing without binary-GPU](https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/12999535417/job/36255045493?pr=1049), is currently failing. Could you please check that?

Thanks, I've fixed it now. I've removed the abstractmethod, because only LQ/LK/LG use it, but not LT. This should fix the issues!

josephleekl commented on .github/workflows/tests_lkcpu_python.yml:
```diff
@@ -243,6 +243,9 @@ jobs:
           OMP_PROC_BIND: false
         run: |
           DEVICENAME=`echo ${{ matrix.pl_backend }} | sed "s/_/./g"`
+          # Temporarily rename source files here to ensure pytest collect source coverage from wheel package
```
IGNORE CHANGES TO THIS FILE

This is to get code coverage for lkcpu, which is from this PR https://github.com/PennyLaneAI/pennylane-lightning/pull/1053/files . This update is temporary
